### PR TITLE
Improvements to CLI to support common workflows

### PR DESCRIPTION
**Pull request by lstein**
# Autocompletion (tab complete) enhancements in dream.py:

- Each of the switches that read a filepath (e.g. --init_img) will trigger file path completion, using either filenames from the current outdir, or using a full or partial path
- The -S switch will display a list of recently-used seeds.

# New commands using the `!command` syntax

In addition to the standard image generation arguments, there are a
series of convenience commands that begin with !:

## !fix

This command runs a post-processor on a previously-generated image. It
takes a PNG filename or path and applies your choice of the -U, -G, or
--embiggen switches in order to fix faces or upscale. If you provide a
filename, the script will look for it in the current output
directory. Otherwise you can provide a full or partial path to the
desired file.

Some examples:

Upscale to 4X its original size and fix faces using codeformer:
~~~
dream> !fix 0000045.4829112.png -G1 -U4 -ft codeformer
~~~

Use the GFPGAN algorithm to fix faces, then upscale to 3X using --embiggen:

~~~
dream> !fix 0000045.4829112.png -G0.8 -ft gfpgan
>> fixing outputs/img-samples/0000045.4829112.png
>> retrieved seed 4829112 and prompt "boy enjoying a banana split"
>> GFPGAN - Restoring Faces for image seed:4829112
Outputs:
[1] outputs/img-samples/000017.4829112.gfpgan-00.png: !fix "outputs/img-samples/0000045.4829112.png" -s 50 -S  -W 512 -H 512 -C 7.5 -A k_lms -G 0.8

dream> !fix 000017.4829112.gfpgan-00.png --embiggen 3
...lots of text...
Outputs:
[2] outputs/img-samples/000018.2273800735.embiggen-00.png: !fix "outputs/img-samples/000017.243781548.gfpgan-00.png" -s 50 -S 2273800735 -W 512 -H 512 -C 7.5 -A k_lms --embiggen 3.0 0.75 0.25
~~~
## !fetch

This command retrieves the generation parameters from a previously
generated image and either loads them into the command line
(Linux|Mac), or prints them out in a comment for copy-and-paste
(Windows). You may provide either the name of a file in the current
output directory, or a full file path.

~~~
dream> !fetch 0000015.8929913.png
# the script returns the next line, ready for editing and running:
dream> a fantastic alien landscape -W 576 -H 512 -s 60 -A plms -C 7.5
~~~

Note that this command may behave unexpectedly if given a PNG file that
was not generated by InvokeAI.

## !history

The dream script keeps track of all the commands you issue during a
session, allowing you to re-run them. On Mac and Linux systems, it
also writes the command-line history out to disk, giving you access to
the most recent 1000 commands issued.

The `!history` command will return a numbered list of all the commands
issued during the session (Windows), or the most recent 1000 commands
(Mac|Linux). You can then repeat a command by using the command !NNN,
where "NNN" is the history line number. For example:

~~~
dream> !history
...
[14] happy woman sitting under tree wearing broad hat and flowing garment
[15] beautiful woman sitting under tree wearing broad hat and flowing garment
[18] beautiful woman sitting under tree wearing broad hat and flowing garment -v0.2 -n6
[20] watercolor of beautiful woman sitting under tree wearing broad hat and flowing garment -v0.2 -n6 -S2878767194
[21] surrealist painting of beautiful woman sitting under tree wearing broad hat and flowing garment -v0.2 -n6 -S2878767194
...
dream> !20
dream> watercolor of beautiful woman sitting under tree wearing broad hat and flowing garment -v0.2 -n6 -S2878767194
~~~

# Other improvements
- normalized how filenames are written out when postprocessing invoked
- various fixes of bugs encountered during testing
- updated documentation
- updated help text
- fix crash that occurs in write_log when user interrupts generation with ^C

### CODE REVIEW

**Comment by tildebyte** [comment]
@lstein;

EXCITEMENT! I finally cracked the 'readline' thing - turns out that simply pip-installing 'pyreadline3' on Windows automagically makes it work.

`import readline` picks up the pyreadline3 implementation, and the APIs are identical.

**Review by tildebyte — APPROVED** [review]
LGTM

**Comment by lstein** [comment]
> EXCITEMENT! I finally cracked the 'readline' thing - turns out that simply pip-installing 'pyreadline3' on Windows automagically makes it work.

Nawwww. Seriously?! This is amazing news. It means we have to revise the documentation, but so worth it!

So does all the tab completion stuff just work? I find it so useful on Linux and it was so frustrating to know that Windows users were missing out.

**Comment by blessedcoolant** [comment]
> @lstein;
> 
> EXCITEMENT! I finally cracked the 'readline' thing - turns out that simply pip-installing 'pyreadline3' on Windows automagically makes it work.
> 
> `import readline` picks up the pyreadline3 implementation, and the APIs are identical.
> 
> EDIT: obviously the "guard" code needs to stay, for insurance

Not working for me on Windows. Maybe I'm missing something else?

**Comment by tildebyte** [comment]
@lstein;

What is https://github.com/invoke-ai/InvokeAI/blob/30c169ea3ba414ba7036ffca97d21ae245758d19/ldm/dream/readline.py#L256 for? pyreadline3 doesn't implement `set_auto_history`; commenting it out lets us proceed to the REPL...

**Comment by tildebyte** [comment]
@blessedcoolant;

- You've pip installed pyreadline3?
- What error (if any) are you getting?

**Comment by tildebyte** [comment]
@blessedcoolant;

Two things occur to me

- I'm launching *without* winpty (should not be needed)
- I'm using git bash, NOT cmd or Powershell or any of the others

**Comment by blessedcoolant** [comment]
> for? pyreadline3 doesn't implement set_auto_history; commenting it out lets us proceed to the REPL...

This is my error. Can confirm. Commenting the line lets me load.

**Comment by tildebyte** [comment]
FWIW, quick testing shows that pyreadline3 on Windows works *without* winpty, under cmd and Powershell as well.

**Comment by tildebyte** [comment]
@lstein ;

Pretty sure that this is an outright bug. Using `!fetch` brings in the text prompt with escaped quotes, which changes the generation:

(line breaks added for readability)

```
a black and white medium format 85mm portrait photograph of
a kitten wearing a tuxedo on his way to a funeral, The photo is
high quality and highly detailed with the kitten\'s features clearly
visible, photographer Edward Weston used Agfa Isopan ISO 25
film to create this image, this image resembles Edward Weston\'s
photograph Pepper No. 35
```


VS. the original

```
a black and white medium format 85mm portrait photograph of
a kitten wearing a tuxedo on his way to a funeral, The photo is
high quality and highly detailed with the kitten's features clearly
visible, photographer Edward Weston used Agfa Isopan ISO 25
film to create this image, this image resembles Edward Weston's
photograph Pepper No. 35
```



(I just now regenerated the "original" version)

All other parameters are constant - there are no other differences in the prompt

**Comment by blessedcoolant** [comment]
> > for? pyreadline3 doesn't implement set_auto_history; commenting it out lets us proceed to the REPL...
> 
> This is my error. Can confirm. Commenting the line lets me load.
> 
> # Suggestions
> * Using `!history` logs !history into the history. I think that's a bit redundant. Is there a way to skip that? Make `!history` not be logged into history.
> * Make commands starting with `!` not be sent to SD. Typing !his accidentally starts processing that with SD. Which in 99.99% of the cases a mistake.
> * Pressing Tab doesn't show !history as one of the available commands. It probably should.
> * Can we save the `dream_history` file in the outputs folder or in the sd folder rather than in the users folder?
> 
> # Bugs Found
> * Trying to `!fetch` a non existing image completely crashes the app. There needs to be a failsafe for this rather than a full crash.
> 
> @lstein

@lstein Just wondering if you'll be addressing these changes?

**Comment by blessedcoolant** [comment]
A reminder that the script is broken on Windows if auto_history line is not disabled.

**Comment by lstein** [comment]
`set_auto_history False` prevents readline from automatically adding every
command to the history list. Without it you can select which commands get
added to the history and which don't. I added it only to prevent the
history from filling up with `!history` commands.

I'll check to see if `pyreadline3` is loaded, and if it is will not set
that option.

Lincoln

**Comment by blessedcoolant** [comment]
Thanks. Will keep an eye out for that PR. : )

**Comment by lstein** [comment]
I created a new PR #870 . There were a number of tricky changes needed to accommodate putting .dream_history into the output dir and I hope I did them right. Otherwise pretty confident that the rest of the bug reports are covered.

How do we handle the requirement for pyreadline3 on Windows? If we import it into the Linux environment, I worry that the auto_save_history() command will disappear. The effect of this is that `!history` commands will start popping up in the history.

**Comment by tildebyte** [comment]
pip and pipenv, at least, can install packages per-platform. Dunno about conda.

**Comment by tildebyte** [comment]
TL;DR - I'm not sure we even have to worry about pyreadline3 conflicts (on Linux, at least)

@lstein; Weird one. I figured a quick, valid test would be trying to `pip install pyreadline3` in a Colab notebook.

OK, that succeeds, surprisingly...

More surprisingly

```python
import readline
readline.set_auto_history(False) 
```

does NOT Except - it Just Works.

Maybe pyreadline3 falls back to built-in on non-Windows...
